tracer/src/Datadog.Trace/AppSec/EventTrackingSdk.cs

```diff
@@ -17,172 +18,185 @@ namespace Datadog.Trace.AppSec;
+    {
+        if (string.IsNullOrEmpty(userId))
         {
-            TelemetryFactory.Metrics.Record(PublicApiUsage.EventTrackingSdk_TrackUserLoginFailureEvent);
-            TrackUserLoginFailureEvent(userId, exists, null, Tracer.Instance);
+            ThrowHelper.ThrowArgumentNullException(nameof(userId));
         }
 
-        /// <summary>
-        /// Sets the details of a logon failure on the local root span
-        /// </summary>
-        /// <param name="userId">The userId associated with the login failure</param>
-        /// <param name="exists">If the userId associated with the login failure exists</param>
-        /// <param name="metadata">Metadata associated with the login failure</param>
-        [PublicApi]
-        public static void TrackUserLoginFailureEvent(string userId, bool exists, IDictionary<string, string> metadata)
-        {
-            TelemetryFactory.Metrics.Record(PublicApiUsage.EventTrackingSdk_TrackUserLoginFailureEvent_Metadata);
-            TrackUserLoginFailureEvent(userId, exists, metadata, Tracer.Instance);
-        }
+        var span = tracer?.ActiveScope?.Span;
 
-        internal static void TrackUserLoginFailureEvent(string userId, bool exists, IDictionary<string, string> metadata, Tracer tracer)
+        if (span is null)
         {
-            if (string.IsNullOrEmpty(userId))
-            {
-                ThrowHelper.ThrowArgumentNullException(nameof(userId));
-            }
-
-            var span = tracer?.ActiveScope?.Span;
-
-            if (span is null)
-            {
-                ThrowHelper.ThrowException("Can't create a tracking event with no active span");
-            }
+            ThrowHelper.ThrowException("Can't create a tracking event with no active span");
+        }
 
-            var setTag = TaggingUtils.GetSpanSetter(span);
+        var setTag = TaggingUtils.GetSpanSetter(span, out var spanInternal);
 
-            setTag(Tags.AppSec.EventsUsers.LoginEvent.FailureTrack, "true");
-            setTag(Tags.AppSec.EventsUsers.LoginEvent.FailureSdkSource, "true");
-            setTag(Tags.AppSec.EventsUsers.LoginEvent.FailureUserId, userId);
-            setTag(Tags.AppSec.EventsUsers.LoginEvent.FailureUserExists, exists ? "true" : "false");
+        setTag(Tags.AppSec.EventsUsers.LoginEvent.FailureTrack, "true");
+        setTag(Tags.AppSec.EventsUsers.LoginEvent.FailureSdkSource, "true");
+        setTag(Tags.AppSec.EventsUsers.LoginEvent.FailureUserId, userId);
+        setTag(Tags.AppSec.EventsUsers.LoginEvent.FailureUserExists, exists ? "true" : "false");
 
-            if (metadata != null)
-            {
-                foreach (var kvp in metadata)
-                {
-                    setTag($"{Tags.AppSec.EventsUsers.LoginEvent.Failure}.{kvp.Key}", kvp.Value);
-                }
-            }
-
-            if (span is Span internalSpan)
+        if (metadata != null)
```

Comment:
Small performance boost opportunity?

   if (metadata is { Count: > 0 })
   {
       var successTagPrefix = $"{Tags.AppSec.EventsUsers.LoginEvent.Success}.";
       foreach (var kvp in metadata)
       {
           setTag(successTagPrefix + kvp.Key, kvp.Value);
       }
   }

Reply:
good idea, changed in [5eb2354](https://github.com/DataDog/dd-trace-dotnet/pull/6225/commits/5eb235421602d965bfe3b2ceb925b15fac34cad7)